### PR TITLE
Add Box impl

### PR DESCRIPTION
**Pull request by ava57r**

### CODE REVIEW

**Review by Enet4 — COMMENTED** [review]
I am not opposed to this, but the native index objects are already heap allocated and thus already "sort of" boxed. With this we are mostly creating a second dereferencing step. Is there a use case for this?

**Comment by ava57r** [comment]
For erase type of index.

Rust std.
https://doc.rust-lang.org/beta/std/boxed/struct.Box.html#impl-Iterator

For code:

```rust 
let mut v: Vec<Box<dyn faiss::Index>> = vec![];

let f1 = FlatIndexImpl::new();
v.push(f1);

let f2 = IndexImpl::new();
v.push(f2);
```

File: src/index/mod.rs [hunk 1]
```diff
@@ -186,12 +186,79 @@ pub trait Index {
     fn set_verbose(&mut self, value: bool);
 }
 
+impl<I> Index for Box<I>
+where
+    I: Index,
+{
+    fn is_trained(&self) -> bool {
+        self.as_ref().is_trained()
```
[the remark this answers is not in the document]
Comment (ava57r) [reply src/index/mod.rs:194]: Yes!. You are right. ;-)

**Comment by Enet4** [comment]
What if we also had a way to upcast a specific index type back to an `IndexImpl`? 

```rust
let f1 = FlatIndex::new_l2(128);
let f2 = index_factory(128, "Flat", MetricType::L2);
let v: Vec<faiss::IndexImpl> = vec![
    f1.upcast(),
    f2,
];
```

**Review by Enet4 — APPROVED** [review]
Looks OK to me nevertheless.

**Comment by ava57r** [comment]
> What if we also had a way to upcast a specific index type back to an `IndexImpl`?
> 
> ```rust
> let f1 = FlatIndex::new_l2(128);
> let f2 = index_factory(128, "Flat", MetricType::L2);
> let v: Vec<faiss::IndexImpl> = vec![
>     f1.upcast(),
>     f2,
> ];
> ```

I agree.

`Box<dyn faiss::Index>` is Rust idiomatic path.

**Comment by ava57r** [comment]
> What if we also had a way to upcast a specific index type back to an `IndexImpl`?
> 
> ```rust
> let f1 = FlatIndex::new_l2(128);
> let f2 = index_factory(128, "Flat", MetricType::L2);
> let v: Vec<faiss::IndexImpl> = vec![
>     f1.upcast(),
>     f2,
> ];
> ```

Do `upcast` in C_API?

**Review by Enet4 — APPROVED** [review]

**Comment by Enet4** [comment]
> Do upcast in C_API?

I don't imagine this to be necessary. The operation is a `reinterpret_cast` between two index pointers, which is a no-op and can be done in Rust. Having functions for a downcast was important because then we could `dynamic_cast`, which checks at run-time that the target type is acceptable.

**Pull request merged**